lib/ansible/modules/cloud/docker/docker_node.py

```diff
@@ -0,0 +1,290 @@
+    pass
+
+from ansible.module_utils.docker_common import (
+    DockerBaseClass,
+)
+
+from ansible.module_utils._text import to_native
+
+from ansible.module_utils.docker_swarm import AnsibleDockerSwarmClient
+
+
+class TaskParameters(DockerBaseClass):
+    def __init__(self, client):
+        super(TaskParameters, self).__init__()
+
+        self.state = None
+
+        # Spec
+        self.name = None
+        self.labels = None
+        self.labels_state = None
+
+        # Node
+        self.availability = None
+        self.role = None
+
+        for key, value in client.module.params.items():
+            setattr(self, key, value)
+
+
+class SwarmNodeManager(DockerBaseClass):
+
+    def __init__(self, client, results):
+
+        super(SwarmNodeManager, self).__init__()
+
+        self.client = client
+        self.results = results
+        self.check_mode = self.client.check_mode
+
+        self.parameters = TaskParameters(client)
+
+    def __call__(self):
+        choice_map = {
+            "update": self.node_update,
+        }
+
+        choice_map.get(self.parameters.state)()
+
+    def node_update(self):
+        if not (self.client.check_if_swarm_manager()):
+            self.client.fail(msg="This node is not a manager.")
+
+        if not (self.client.check_if_swarm_node(node_id=self.parameters.name)):
+            self.results['actions'].append("This node is not part of a swarm.")
+            return
+
+        try:
+            status_down = self.client.check_if_swarm_node_is_down()
+        except APIError:
+            return
+
+        if status_down:
+            self.client.fail(msg="Can not update the node. The status node is down.")
+
+        try:
+            node_info = self.client.inspect_node(node_id=self.parameters.name)
+        except APIError as exc:
+            self.client.fail(msg="Failed to get node information for %s" % to_native(exc))
+
+        if self.parameters.role is None:
+            self.parameters.role = node_info['Spec']['Role']
+
+        if self.parameters.availability is None:
+            self.parameters.availability = node_info['Spec']['Availability']
+
+        if self.parameters.labels is None:
+            self.parameters.labels = node_info['Spec']['Labels']
+        else:
+            if self.parameters.labels_state == 'remove':
+                self.parameters.labels = {}
+            if self.parameters.labels_state == 'replace':
+                pass
+            if self.parameters.labels_state == 'merge':
+                for next_key in node_info['Spec']['Labels']:
+                    if next_key not in self.parameters.labels:
+                        self.parameters.labels.update({next_key: node_info['Spec']['Labels'][next_key]})
+
+        node_spec = dict(
+            Availability=self.parameters.availability,
+            Role=self.parameters.role,
+            Labels=self.parameters.labels,
+        )
+
+        try:
```

Comment:
Where do you determine that the node's state actually needs to be updated?

Reply:
I don't think it is necessary because it is up to the internal Docker engine to decide what to do. No matter if you are changing something or not Swarm manager exchanges packets with node. Neither CLI or API informs you that you did not change anything. And I don't see the necessity to add it here, but if we do there should be `force` mode to send the update even if nothing changes.

The `Version`, `Availability` and `Role` are always required, if the `Label` is missing, then labels are removed from the node.

Reply:
It is customary that Ansible modules return `changed == False` if nothing has to be changed. So you have to determine it anyway. And not doing an API call when you already know that it is not needed is better anyway.

Reply:
This is not a problem, it has to inspect node before making changes anyway to get up-to-date `Version`

Reply:
Change tracking added in commit 9e54a068